src/Common/src/Interop/Windows/perfcounter/Interop.PerformanceData.cs

```diff
@@ -0,0 +1,93 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.Runtime.InteropServices;
+using Microsoft.Win32.SafeHandles;
+
+internal static partial class Interop
+{
+    internal partial class PerfCounter
+    {        
+        [DllImport(Libraries.Advapi32, ExactSpelling = true)]
+        internal static extern unsafe uint PerfStopProvider(
+            IntPtr hProvider
+        );
+
+        internal unsafe delegate uint PERFLIBREQUEST(
+            uint RequestCode,
+            void* Buffer,
+            uint BufferSize
+        );
+
+        // Native PERFLIB V2 Provider APIs.
+        internal struct PerfCounterSetInfoStruct
+        {
+            // PERF_COUNTERSET_INFO structure defined in perflib.h
+            internal Guid CounterSetGuid;
+            internal Guid ProviderGuid;
+            internal uint NumCounters;
+            internal uint InstanceType;
+        }
+
+        internal struct PerfCounterInfoStruct
+        {
+            // PERF_COUNTER_INFO structure defined in perflib.h
+            internal uint CounterId;
+            internal uint CounterType;
+            internal long Attrib;
+            internal uint Size;
+            internal uint DetailLevel;
+            internal uint Scale;
+            internal uint Offset;
+        }
+
+        [StructLayout(LayoutKind.Explicit, Size = 32)]
+        internal struct PerfCounterSetInstanceStruct
+        {
+            // PERF_COUNTERSET_INSTANCE structure defined in perflib.h
+            [FieldOffset(0)] internal Guid CounterSetGuid;
+            [FieldOffset(16)] internal uint dwSize;
+            [FieldOffset(20)] internal uint InstanceId;
+            [FieldOffset(24)] internal uint InstanceNameOffset;
+            [FieldOffset(28)] internal uint InstanceNameSize;
+        }
+
+        [DllImport(Libraries.Advapi32, ExactSpelling = true)]
+        internal static extern unsafe uint PerfStartProvider(
+            ref Guid ProviderGuid,
+            PERFLIBREQUEST ControlCallback,
+            out SafePerfProviderHandle phProvider
+        );
+
+        [DllImport(Libraries.Advapi32, SetLastError = true, ExactSpelling = true, CharSet = CharSet.Unicode)]
+        internal static extern unsafe PerfCounterSetInstanceStruct* PerfCreateInstance(
+            SafePerfProviderHandle hProvider,
+            ref Guid CounterSetGuid,
+            string szInstanceName,
+            uint dwInstance
+        );
+
+        [DllImport(Libraries.Advapi32, ExactSpelling = true)]
+        internal static extern unsafe uint PerfSetCounterSetInfo(
+            SafePerfProviderHandle hProvider,
+            PerfCounterSetInfoStruct* pTemplate,
```

Comment:
i removed `[In][Out]` guideline says to use `ref` but on callsite i cannot pass a pointer to `ref`. Can you help me to understand how to resolve, i'm not so "fluent" with pinvoke.
/cc @jkotas @Anipik @JeremyKuhne

Reply:
NB. It works, maybe i don't need to use `ref` here.

Reply:
Yeah sometimes we can get away with using the ```ref```